### DISCUSSION TITLE
Creating a custom directive

### DISCUSSION THREAD

**Discussion by personalsoft-fabianobonin** [alpinejs/alpine#1815] [1. Help] [answered]
Hi,

I need to bind to element properties instead of attributes, and the only way I found to achieve this task was creating a custom directive called `x-prop`

Let's say my element have a custom property called `myprop`

```html
<body x-data="data"
  ps-init>
  <my-element x-prop:myprop="obj"></my-element>
  <script>
    const data = {
      obj: {
        key1: 'value1',
        key2: 'value2'
      }
    }
  </script>
</body>
```

I wrote the code below for my custom directive.
I am not used to Alpine project, so this is basically a copy/paste of code fragments from `directives/x-bind.js`

```typescript
Alpine.directive('prop', (el: any, { value, expression }: any, { effect }: any) => {
  const evaluate = Alpine.evaluateLater(el, expression);

  effect(() => evaluate((result: any) => {
    if (!el._x_bindings) el._x_bindings = Alpine.reactive({});
    el._x_bindings[value] = result;

    Alpine.mutateDom(() => {
      el[value] = result;
    });
  }));
});
```

The code is working, but when it evaluates the expression, the `result` argument receives a proxy to `data.obj`, so the proxy is stored in `<my-element>.myprop` instead of the original `data.obj`

Is there a way I can access the proxy's target from the `evaluate` function?

Fabiano

**Comment by KevinBatdorf** [comment, marked as answer]
I think you want to use `closestRoot(el)` instead of `el`. If you get stuck and put it on codepen or elsewhere I can show you

**Reply by personalsoft-fabianobonin** [reply]
The related problem is not in `el`, as I am receiving the expected element in this variable.
The problem is in the `result` argument, which is a proxy to `data.obj`.
I need to access the underlying `data.obj`.

**Reply by KevinBatdorf** [reply]
You can treat JS proxies like normal objects. If you need to unwrap the proxy, Alpine has a `raw` method too, but I don't think that's what you want. 

If you have it online somewhere I can tweak it for you.

**Reply by personalsoft-fabianobonin** [reply]
I tried on codepen (https://codepen.io/fabianobonin/pen/WNjEXBX), but codepen console probably stringfy the result of `console.log`.

You can check there is a proxy in the element's property with:

`document.querySelector('my-element').myprop`

```html
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <script defer
      src="https://unpkg.com/alpinejs@3.x.x/dist/cdn.min.js"></script>
  </head>
  <body x-data="data"
    ps-init>
    <my-element x-prop:myprop="obj"></my-element>
    <script>
      const data = {
        obj: {
          key1: 'value1',
          key2: 'value2'
        }
      }

      document.addEventListener('alpine:init', () => {
        Alpine.directive('prop', (el, { value, expression }, { effect }) => {
          const evaluate = Alpine.evaluateLater(el, expression);

          effect(() => evaluate((result) => {
            Alpine.mutateDom(() => {
              el[value] = result;
            });
          }));
        });
      });
    </script>
  </body>
</html>
```

**Reply by KevinBatdorf** [reply]
Okay, so yeah as I mentioned, you should be able to use the proxy as you would any object. But if you want to remove the reactivity, you can unwrap it using `raw()`

```js
Alpine.directive('prop', (el, { value, expression }, { Alpine: { raw }, effect })
[...]
el[value] = raw(result)
```

Is that what you want?

**Reply by personalsoft-fabianobonin** [reply]
Kevin, aparently `Alpine.raw()` was what I needed.

Now it is storing the original object instead of the proxy.

```javascript
Alpine.mutateDom(() => {
  el[value] = Alpine.raw(result);
});
```

Thanks for your support!

Fabiano

**Reply by personalsoft-fabianobonin** [reply]
Wrote my message at the same time as you...

Thanks again, I will check if I really need to use the raw object.

My custom property is an array of objects, and each time the user adds an element to the array, Alpine is creating a proxy over a proxy and so on... So at the end, if I add 10 elements, I have a proxy with 10 levels on the first element.

Best regards,

Fabiano

**Reply by personalsoft-fabianobonin** [reply]
Here is the final code:

```javascript
    // Custom 'x-prop' directive for property binding ('x-bind' binds to attributes)
    Alpine.directive('prop', (el, { value, expression }, { Alpine: { evaluateLater, mutateDom, raw }, effect }) => {
      const evaluate = evaluateLater(el, expression);

      effect(() => evaluate((result) => {
        mutateDom(() => {
          // The value property on <input> element doesn't accept null or undefined
          if (result == null && el instanceof HTMLInputElement) {
            el.value = '';
          } else {
            el[value] = raw(result);
          }
        });
      }));
    });
```
